### PR TITLE
CA2213:Disposable fields should be disposed and support for `DisposeCoreAsync`

### LINKED ISSUE #4950: CA2213 does not recognize an IAsyncDisposable implementaton as sufficient.
### Analyzer

**Diagnostic ID**: [CA2013](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2013): `Do not use ReferenceEquals with value types`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

### Describe the bug

CA2213 seemingly does not consider [DisposeAsyncCore](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#disposeasync-and-disposeasynccore) method. 

```csharp
/// <summary>State object for <see cref="RestClient"/>.</summary>
private readonly RestClientContext restContext; // <<<< CA2213: WARNING EMMITED HERE.

// ...

/// <summary>
/// Frees managed resources used by the object.
/// </summary>
/// <returns>A <see cref="ValueTask">task</see> that represents the asynchronous dispose operation.</returns>
protected virtual async ValueTask DisposeAsyncCore()
{
    this.log.Debug("*");

    lock (this.disposedValueLock)
    {
        if (this.disposedValue)
        {
            this.log.Debug("$<ALREADY_DISPOSED>");
            return;
        }

        this.disposedValue = true;
    }

    this.log.Debug("Dispose REST context.");
    this.restContext.Dispose(); // // This does not make the warning go away!

    this.log.Debug("$");
}

/// <inheritdoc/>
public async ValueTask DisposeAsync()
{
    // this.restContext.Dispose(); // THIS MAKES THE WARNING GO AWAY.
    await this.DisposeAsyncCore().ConfigureAwait(false);
    GC.SuppressFinalize(this);
}
```

### Steps To Reproduce

Please see the source code above.

### Expected behavior

No warning is emitted when `this.restContext.Dispose();` is called in `DisposeAsyncCore`.

### Actual behavior

A warning is emitted when `this.restContext.Dispose();` is called in `DisposeAsyncCore`.

### PR DESCRIPTION
**Pull request by MartyIX**
Attempts to fix #4950 but I don't have any clue if the approach of the PR is correct or not.

However, this seems to be in line with https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-disposeasync#implement-the-async-dispose-pattern except for naming `DisposeCoreAsync` vs `DisposeAsyncCore`:

```cs
public class ExampleAsyncDisposable : IAsyncDisposable
{
    private IAsyncDisposable? _example;

    public ExampleAsyncDisposable() =>
        _example = new NoopAsyncDisposable();

    public async ValueTask DisposeAsync()
    {
        await DisposeAsyncCore().ConfigureAwait(false);

        GC.SuppressFinalize(this);
    }

    protected virtual async ValueTask DisposeAsyncCore()
    {
        if (_example is not null)
        {
            await _example.DisposeAsync().ConfigureAwait(false);
        }

        _example = null;
    }
}
```

Somewhat unrelated: I created a PR in [vs-threading](https://github.com/microsoft/vs-threading) repo where I [claimed](https://github.com/microsoft/vs-threading/pull/1238#issue-1930821499) that `DisposeAsyncCore` seemed to be the correct name for the method but this repo uses `DisposeCoreAsync` (AFAIK). It's confusing.

### CODE REVIEW

File: src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs [hunk 1]
```diff
@@ -300,6 +300,18 @@ private static bool HasOverriddenDisposeCoreAsyncMethodSignature(this IMethodSym
                 method.Parameters[0].Type.SpecialType == SpecialType.System_Boolean;
         }
 
+        /// <summary>
+        /// Checks if the given method has the signature "virtual ValueTask DisposeCoreAsync()".
+        /// </summary>
+        private static bool HasVirtualDisposeCoreAsyncMethodSignature(this IMethodSymbol method, [NotNullWhen(returnValue: true)] INamedTypeSymbol? task)
```
Comment (mavasani) [inline src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs:306]:
```suggestion
        private static bool HasVirtualDisposeCoreAsyncMethodSignature(this IMethodSymbol method, [NotNullWhen(returnValue: true)] INamedTypeSymbol? valueTask)
```

File: src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs [hunk 2]
```diff
@@ -300,6 +300,18 @@ private static bool HasOverriddenDisposeCoreAsyncMethodSignature(this IMethodSym
...
+        {
+            return method.Name == "DisposeCoreAsync" &&
+                method.MethodKind == MethodKind.Ordinary &&
+                method.IsVirtual &&
+                SymbolEqualityComparer.Default.Equals(method.ReturnType, task) &&
```
Comment (mavasani) [inline src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs:311]:
```suggestion
                SymbolEqualityComparer.Default.Equals(method.ReturnType, valueTask) &&
```

**Comment by mavasani** [comment]
Implementation of the fix looks fine to me, but I am not familiar with the semantics and naming requirements of the API for recommended dispose patterns. I'll let someone from the runtime team comment on it - @buyaa-n @carlossanlop @stephentoub @jeffhandley

**Comment by buyaa-n** [comment]
Sorry for late reply @MartyIX, the latest `Framework Design Guidelines` suggests the name `DisposeAsyncCore`, so we should use `DisposeAsyncCore`.

I have never seen `DisposeCoreAsync` usage, though got 2 hits [from source.net search](https://source.dot.net/#q=DisposeCoreAsync), seems it is rare but used, probably only in old code. Not sure if we should cover both names.

**Comment by buyaa-n** [comment]
Thanks @MartyIX, definitely we need to support both names. Looks we could just adjust the name check in `HasOverriddenDisposeCoreAsyncMethodSignature` and `HasVirtualDisposeCoreAsyncMethodSignature`, plus update the doc/comments that it covers both versions. Doesn't not necessarily need to change the method name and Enum member name `DisposeCoreAsync`, what you think? I'll leave that up to you and @mavasani.

**Comment by mavasani** [comment]
> Thanks @MartyIX, definitely we need to support both names. Looks we could just adjust the name check in `HasOverriddenDisposeCoreAsyncMethodSignature` and `HasVirtualDisposeCoreAsyncMethodSignature`, plus update the doc/comments that it covers both versions. Doesn't not necessarily need to change the method name and Enum member name `DisposeCoreAsync`, what you think? I'll leave that up to you and @mavasani.

Agree with all of @buyaa-n's assessment.

**Comment by paulomorgado** [comment]
> @mavasani @buyaa-n Could you please check the latest version of the PR?
> 
> I'm having difficulty with the VB test but then I'm not sure if it is required at all.

What VB code are you having issues with?

**Comment by paulomorgado** [comment]
> > > @mavasani @buyaa-n Could you please check the latest version of the PR?
> > > I'm having difficulty with the VB test but then I'm not sure if it is required at all.
> > 
> > 
> > What VB code are you having issues with?
> 
> Please see [#6976 (comment)](https://github.com/dotnet/roslyn-analyzers/pull/6976#discussion_r1384757344). It's commented out. I can't figure out how to rewrite C# code above the commented out VB code.

It would help if you said what particular line of code is getting that error.

Looks like **Visual Basic**'s [Asynchronous programming with Async and Await](https://learn.microsoft.com/en-us/dotnet/visual-basic/programming-guide/concepts/async/) does not support `ValueTask`.

**Review by mavasani — APPROVED** [review]

**Comment by mavasani** [comment]
@buyaa-n Please review the PR and merge if it looks good to you.

File: src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs [hunk 3]
```diff
@@ -288,18 +288,30 @@ private static bool HasDisposeCloseAsyncMethodSignature(this IMethodSymbol metho
         }
 
         /// <summary>
-        /// Checks if the given method has the signature "override Task DisposeCoreAsync(bool)".
+        /// Checks if the given method has the signature "override Task DisposeCoreAsync(bool)" or "override Task DisposeAsyncCore(bool)".
         /// </summary>
         private static bool HasOverriddenDisposeCoreAsyncMethodSignature(this IMethodSymbol method, [NotNullWhen(returnValue: true)] INamedTypeSymbol? task)
         {
-            return method.Name == "DisposeCoreAsync" &&
+            return (method.Name == "DisposeCoreAsync" || method.Name == "DisposeAsyncCore") &&
```
Comment (buyaa-n) [inline src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs:295]: I would check the `DisposeAsyncCore` first, as it is the mostly used version
```suggestion
            return (method.Name == "DisposeAsyncCore" || method.Name == "DisposeCoreAsync") &&
```

File: src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs [hunk 4]
```diff
@@ -288,18 +288,30 @@ private static bool HasDisposeCloseAsyncMethodSignature(this IMethodSymbol metho
...
                 method.Parameters.Length == 1 &&
                 method.Parameters[0].Type.SpecialType == SpecialType.System_Boolean;
         }
 
+        /// <summary>
+        /// Checks if the given method has the signature "virtual ValueTask DisposeCoreAsync()" or "virtual ValueTask DisposeAsyncCore()".
+        /// </summary>
+        private static bool HasVirtualDisposeCoreAsyncMethodSignature(this IMethodSymbol method, [NotNullWhen(returnValue: true)] INamedTypeSymbol? valueTask)
+        {
+            return (method.Name == "DisposeCoreAsync" || method.Name == "DisposeAsyncCore") &&
```
Comment (buyaa-n) [inline src/Utilities/Compiler/Extensions/IMethodSymbolExtensions.cs:308]: NIT: the same
```suggestion
            return (method.Name == "DisposeAsyncCore" || method.Name == "DisposeCoreAsync") &&
```

**Review by buyaa-n — APPROVED** [review]
LGTM, thank you @MartyIX

**Pull request merged**